File: DSA/GFG/Heap/k_largest_elements.py
# Link: https://www.geeksforgeeks.org/batch/gfg-160-problems/track/heap-gfg-160/problem/k-largest-elements4206

# Python program to find the k largest elements in the  
# array using min heap

import heapq

# Function to find the k largest elements in the array 
def kLargest(arr, k):
  
    # Create a min-heap with the first k elements
    minH = arr[:k]
    heapq.heapify(minH)
    
    # Traverse the rest of the array
    for x in arr[k:]:
        if x > minH[0]:
            heapq.heapreplace(minH, x)
    
    res = []

    # Min heap will contain only k 
    # largest element
    while minH:
        res.append(heapq.heappop(minH))

    # Reverse the result array, so that all
    # elements are in decreasing order
    res.reverse()

    return res

# ============================================================================
#  Driver program to test above functions
if __name__ == "__main__":
    arr = [1, 23, 12, 9, 30, 2, 50]
    k = 3
    res = kLargest(arr, k)
    print(" ".join(map(str, res)))

# Time Complexity: O(n log k)
# Space Complexity: O(k)

File: DSA/GFG/Heap/k_largest_elements.cpp
// Link: https://www.geeksforgeeks.org/batch/gfg-160-problems/track/heap-gfg-160/problem/k-largest-elements4206

// C++ program to find the k largest elements in the array
// using partitioning step of quick sort

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Function to partition the array around a pivot
int partition(vector<int> &arr, int left, int right)
{

    // Last element is chosen as a pivot.
    int pivot = arr[right];
    int i = left;

    for (int j = left; j < right; j++)
    {

        // Elements greater than or equal to pivot
        // are placed in the left side of pivot
        if (arr[j] >= pivot)
        {
            swap(arr[i], arr[j]);
            i++;
        }
    }

    swap(arr[i], arr[right]);

    // The correct sorted position of the pivot
    return i;
}

void quickSelect(vector<int> &arr, int left, int right, int k)
{
    if (left <= right)
    {
        int pivotIdx = partition(arr, left, right);

        // Count of all elements in the left part
        int leftCnt = pivotIdx - left + 1;

        // If leftCnt is equal to k, then we have
        // found the k largest element
        if (leftCnt == k)
            return;

        // Search in the left subarray
        if (leftCnt > k)
            quickSelect(arr, left, pivotIdx - 1, k);

        // Reduce the k by number of elements already covered
        // and search in the right subarray
        else
            quickSelect(arr, pivotIdx + 1, right, k - leftCnt);
    }
}

vector<int> kLargest(vector<int> &arr, int k)
{
    int n = arr.size();
    quickSelect(arr, 0, n - 1, k);

    // First k elements of the array will be the largest
    vector<int> res(arr.begin(), arr.begin() + k);

    // Sort the result in descending order
    sort(res.begin(), res.end(), greater<int>());
    return res;
}

// ============================================================================
// Driver program to test above functions
int main()
{
    vector<int> arr = {1, 23, 12, 9, 30, 2, 50};
    int k = 3;

    vector<int> res = kLargest(arr, k);
    for (int ele : res)
        cout << ele << " ";

    return 0;
}

// Time Complexity: O(n log k)
// Space Complexity: O(k)